django/db/models/functions/datetime.py

```diff
@@ -0,0 +1,229 @@
+DateTimeField.register_lookup(ExtractHour)
+DateTimeField.register_lookup(ExtractMinute)
+DateTimeField.register_lookup(ExtractSecond)
+
+ExtractYear.register_lookup(YearExact)
+ExtractYear.register_lookup(YearGt)
+ExtractYear.register_lookup(YearGte)
+ExtractYear.register_lookup(YearLt)
+ExtractYear.register_lookup(YearLte)
+
+
+class TruncBase(Transform):
+    arity = 1
+    kind = None
+    tzinfo = None
+
+    def __init__(self, expression, output_field=None, tzinfo=None, **extra):
+        self.tzinfo = tzinfo
+        super(TruncBase, self).__init__(expression, output_field=output_field, **extra)
+
+    def as_sql(self, compiler, connection):
+        inner_sql, inner_params = compiler.compile(self.lhs)
+        # Escape any params because trunc_sql will format our string
+        inner_sql = inner_sql.replace('%s', '%%s')
+        if isinstance(self.output_field, DateTimeField):
+            tzname = None
+            if settings.USE_TZ:
+                # Apply the timezone conversion to the datetime *before*
+                # truncation. 2015-12-31 23:00:00 -02:00 is stored in the
+                # database as 2016-01-01 01:00:00 +00:00. Truncating to year
+                # should truncate to 2015-01-01 if the user is in -02:00, not
+                # 2016.
+                if self.tzinfo is None:
+                    tzname = timezone.get_current_timezone_name()
+                else:
+                    tzname = timezone._get_timezone_name(self.tzinfo)
+            sql, params = connection.ops.datetime_trunc_sql(self.kind, inner_sql, tzname)
+        elif isinstance(self.output_field, DateField):
+            sql = connection.ops.date_trunc_sql(self.kind, inner_sql)
+            params = []
+        else:
+            raise ValueError('Trunc only valid on Date or DateTime fields')
+        return sql, inner_params + params
+
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
+        copy = super(TruncBase, self).resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        field = copy.lhs.output_field
+        # DateTimeField is a subclass of DateField so this works for both.
+        assert isinstance(field, DateField), (
+            "%r isn't a DateField or DateTimeField." % field.name
+        )
+        # If self.output_field was None, then accessing the field will trigger
+        # the resolver to assign it to self.lhs.output_field.
+        if not isinstance(copy.output_field, (DateField, DateTimeField)):
+            raise ValueError('output_field must be either DateField or DateTimeField')
+        # Passing dates to functions expecting datetimes is most likely a
+        # mistake.
+        if type(field) == DateField and (
+                isinstance(copy.output_field, DateTimeField) or copy.kind in ('hour', 'minute', 'second')):
+            raise ValueError("Attempting to convert DateTime %r to DateTimeField. " % field.name)
```

Comment:
@aaugustin this is what I came up with to prevent dates being passed to functions expecting datetimes. The error is now a ValueError rather than an AssertionError, but there were no tests present to enforce this for `datetimes()`.

It seemed better to enforce this invariant in Trunc rather than the frontend .datetimes() queryset method. That, and using meta to traverse related fields to query types in the frontend method seemed like overkill 😈 

The obvious problem is that any datetime `kinds` that aren't explicitly supported by django do *not* enforce this restriction. Not a problem for `dates()` and `datetimes()` which do restrict the set of `kinds` available, and django can't know the actual expected type for unsupported kinds anyway.

I think this is a reasonable solution. I'm not entirely happy with the error message so I'm keen to hear advice on that aspect for anyone doing review.